src/main/java/javax/jmdns/impl/util/SimpleLockManager.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright Object Matrix 2017
+ */
+
+package javax.jmdns.impl.util;
+
+import java.io.Closeable;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Basic lock manager which uses internal {@link Map} to maintain a list of current locks.
+ */
+public class SimpleLockManager {
+
+    private final ConcurrentHashMap<String, Locked> _locks = new ConcurrentHashMap<>();
+
+    public Locked lock(String lockKey) {
+        try {
+            return tryLock(lockKey, Long.MAX_VALUE, TimeUnit.MILLISECONDS);
+        } catch (LockFailedException e) {
+            throw new RuntimeException(e);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException(e);
+        }
+    }
+
+    @SuppressWarnings("resource")
+    public Locked tryLock(String lockKey, long time, TimeUnit timeunit) throws InterruptedException, LockFailedException {
```

Comment:
Instead of implementing the mechanism ourselves, I would prefer to use standard Java mechanisms such as `Semaphore`. Or am I missing any reason why this isn't possible here?

Reply:
Hi
Thanks for your feedback

My idea was as follows
- reduce contention to a specific key, rather than a global lock
- provide Lock implementing  Closeable interface to use in try-with-resource and have early compilation warning if not closed. Instead Java's Lock interface would work here

Semaphore is usually used to control number of concurrent threads. Here likely it would be used with number of permits = 1, so a global lock, equivalent to synchronized block
It would still require using something like ConcurrentHashMap for keeping track of key->lock mapping

I'm happy to re-implement it if you can think of a simple pattern I could use here?

Reply:
Sorry for dropping the ball here - thanks for your explanations!